client/src/main/java/org/apache/rocketmq/client/producer/selector/SelectMessageQueueByHash.java

```diff
@@ -25,12 +26,12 @@
 
     @Override
     public MessageQueue select(List<MessageQueue> mqs, Message msg, Object arg) {
-        int value = arg.hashCode();
-        if (value < 0) {
-            value = Math.abs(value);
-        }
+        return mqs.get(indexForQueue(mqs.size(), arg));
+    }
 
-        value = value % mqs.size();
-        return mqs.get(value);
+    private int indexForQueue(int length, Object key) {
+        int h = (key == null) ? 0 : (h = key.hashCode()) ^ (h >>> 16);
+        return h & (length - 1);
```

Comment:
Maybe a bug :-)
When mqs.size() is not power of 2,  some queue will never be selected.
for example. when mqs.size() is 5, only queue-0 and queue-4 could be selected.

Reply:
default of queue nums is 4 ,we recommend it power of 2 , donnot we? we can sign if you want your bussiness message ordered by using 'SelectMessageQueueByHash' , you can config your  queue nums   power of 2 for good performance, like hashmap,if you sign the constructor of hashmap 'HashMap(int initialCapacity) ',  initialCapacity if not be power of 2 ,maybe bug  also ,moreever  ,the code before really perform not balance

Reply:
Yes, for many scenes, setting the queue number to power of 2 is a best practice.  But Rocketmq did not require the queue number of a topic MUST BE power of 2 by appointment. I think is not a good choice to break the contract here. 
For the unbalance case you mentioned, my advice is to check if hotspot data exist.
If you prefer to insist the queue number should be power of 2,  adding a predict check before invoking the Hash algorithm may be a good choice. 
:-)

Reply:
I have accepted your suggestion， only use algorithm  when queue nums is power of 2 by code 'if ((length & (length - 1)) == 0)', otherwise algorithm like before